From a2f79d08b98a06af64b3c2a3555be424baf8703a Mon Sep 17 00:00:00 2001
From: Nicolas Fella <nicolas.fella@gmx.de>
Date: Sun, 9 Feb 2025 16:48:57 +0100
Subject: [PATCH] KeyboardBrightnessController: Drop isValid() check for upower
 interface

isValid() seems to be racy when upower is just starting up (because we
DBus-activated it earlier in BatteryController).

This causes us to think keyboard brightness control is not available even
when it actually is.

To address this always make a call to the interface. The call will
block until the interface is ready to either give us the correct result
or we know it's not available because we get a relevant DBus error.

As a bonus print a warning message if we get any unexpected errors

BUG: 486067
---
 .../keyboardbrightnesscontroller.cpp          | 32 +++++++++----------
 1 file changed, 16 insertions(+), 16 deletions(-)

diff --git a/daemon/controllers/keyboardbrightnesscontroller.cpp b/daemon/controllers/keyboardbrightnesscontroller.cpp
index ff2529866..345b89fdc 100644
--- a/daemon/controllers/keyboardbrightnesscontroller.cpp
+++ b/daemon/controllers/keyboardbrightnesscontroller.cpp
@@ -30,22 +30,22 @@ KeyboardBrightnessController::KeyboardBrightnessController()
 {
     m_kbdBacklight =
         new OrgFreedesktopUPowerKbdBacklightInterface(UPOWER_SERVICE, u"/org/freedesktop/UPower/KbdBacklight"_s, QDBusConnection::systemBus(), this);
-    if (m_kbdBacklight->isValid()) {
-        // Cache max value
-        QDBusPendingReply<int> rep = m_kbdBacklight->GetMaxBrightness();
-        rep.waitForFinished();
-        if (rep.isValid()) {
-            m_maxBrightness = rep.value();
-            m_isSupported = true;
-        }
-        // TODO Do a proper check if the kbd backlight dbus object exists. But that should work for now ..
-        if (m_maxBrightness) {
-            m_cachedBrightness = brightness();
-            qCDebug(POWERDEVIL) << "current keyboard backlight brightness value: " << m_cachedBrightness;
-            connect(m_kbdBacklight,
-                    &OrgFreedesktopUPowerKbdBacklightInterface::BrightnessChangedWithSource,
-                    this,
-                    &KeyboardBrightnessController::onBrightnessChanged);
+
+    QDBusPendingReply<int> rep = m_kbdBacklight->GetMaxBrightness();
+    rep.waitForFinished();
+    if (rep.isValid()) {
+        m_maxBrightness = rep.value();
+        m_isSupported = true;
+        m_cachedBrightness = brightness();
+        qCDebug(POWERDEVIL) << "current keyboard backlight brightness value: " << m_cachedBrightness;
+        connect(m_kbdBacklight,
+                &OrgFreedesktopUPowerKbdBacklightInterface::BrightnessChangedWithSource,
+                this,
+                &KeyboardBrightnessController::onBrightnessChanged);
+    } else {
+        // Don't warn when no keyboard backlight is available, only for other errors
+        if (rep.error().type() != QDBusError::UnknownMethod) {
+            qCWarning(POWERDEVIL) << "Could not query keyboard backlight brightness" << rep.error().message();
         }
     }
 }
-- 
GitLab

